clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -252,11 +305,38 @@ private void setSaslState(SaslState saslState) {
             this.pendingSaslState = null;
             this.saslState = saslState;
             LOG.debug("Set SASL client state to {}", saslState);
-            if (saslState == SaslState.COMPLETE)
-                transportLayer.removeInterestOps(SelectionKey.OP_WRITE);
+            if (saslState == SaslState.COMPLETE) {
+                setAuthenticationEndAndSessionReauthenticationTimes();
+                /*
+                 * Re-authentication is triggered by a write, so we have to make sure that
+                 * pending write is actually sent.
+                 */
+                transportLayer.addInterestOps(SelectionKey.OP_WRITE);
+            }
         }
     }
 
+    private void setAuthenticationEndAndSessionReauthenticationTimes() {
+        authenticationEndMs = time.milliseconds();
+        if (positiveSessionLifetimeMs != null) {
+            // pick a random percentage between 85% and 95% for session re-authentication
+            double pctWindowFactorToTakeNetworkLatencyAndClockDriftIntoAccount = 0.85;
+            double pctWindowJitterToAvoidReauthenticationStormAcrossManyChannelsSimultaneously = 0.10;
+            double pctToUse = pctWindowFactorToTakeNetworkLatencyAndClockDriftIntoAccount + RNG.nextDouble()
+                    * pctWindowJitterToAvoidReauthenticationStormAcrossManyChannelsSimultaneously;
+            clientSessionReauthenticationTimeMs = authenticationEndMs
+                    + (long) (positiveSessionLifetimeMs.longValue() * pctToUse);
```

Comment:
On the broker-side, I think we need a minimum re-authentication interval as well to restrict the rate at which clients re-authenticate. This is particularly important since we dont apply any quotas for authentication. Without imposing a re-authentication rate limit, a client that enters a re-authentication loop (due to a bug or intentionally) would effectively stop the broker from doing anything useful.

Reply:
👍 I defined a 1-second requirement that applies to the second and subsequent re-authentications.  So the first re-authentication can happen immediately after authentication if desired, but the second re-authentication must then happen at least 1 second later (and so on), otherwise the SASL handshake is passed through without beginning the re-authentication process (and that would mean the connection is closed, which also results in the client experience the newly-implemented DDoS delay).  There are two reasons I set it for the second re-auth.  One is is because `SaslAuthenticatorTest` would end up running longer if we had to set the interval that much longer.  The other is that we don't have a time value that we can use to set the start time until `KafkaChannel.maybeBeginServerReauthentication()` is invoked.  I think this is reasonable -- is it okay with you?

Reply:
@rajinisivaram Just want to confirm that this is acceptable.